File: app/src/main/java/com/udacity/maluleque/meutako/TransactionListFragment.java
package com.udacity.maluleque.meutako;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.cardview.widget.CardView;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.ListenerRegistration;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.QueryDocumentSnapshot;
import com.udacity.maluleque.meutako.adapters.TransactionsAdapter;
import com.udacity.maluleque.meutako.model.Transaction;
import com.udacity.maluleque.meutako.preferences.PreferencesManager;
import com.udacity.maluleque.meutako.utils.DateUtils;
import com.udacity.maluleque.meutako.utils.NumberUtils;
import com.udacity.maluleque.meutako.widget.TransactionViewService;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

import static android.content.Context.MODE_PRIVATE;


public class TransactionListFragment extends Fragment implements TransactionsAdapter.OnTransactionClickListener {
    private static final String ARG_PARAM1 = "param1";
    private static final String ARG_PARAM2 = "param2";
    private static final String TAG = "TransactionListFragment";
    private static final String LAUNCH_PREF = "launch-prefs";

    @BindView(R.id.recyclerViewTransactions)
    RecyclerView recyclerView;

    @BindView(R.id.noDataLayout)
    LinearLayout noDataLayout;

    @BindView(R.id.noInternetLayout)
    LinearLayout noInternetLayout;

    @BindView(R.id.progressBar)
    ProgressBar progressBar;

    @BindView(R.id.textViewAvailableAmount)
    TextView textViewAvailable;
    @BindView(R.id.textViewIncome)
    TextView textViewIncome;
    @BindView(R.id.textViewExpense)
    TextView textViewExpense;

    @BindView(R.id.resumeCardView)
    CardView resumeCardView;



    private int mParam1;
    private String dataMonth;
    private FirebaseFirestore db;
    private ListenerRegistration registration;
    private FirebaseUser user;
    private FabButtonVisibilityListener fabButtonVisibilityListener;

    public TransactionListFragment() {

    }


    public static TransactionListFragment newInstance(int param1, String param2) {
        TransactionListFragment fragment = new TransactionListFragment();
        Bundle args = new Bundle();
        args.putInt(ARG_PARAM1, param1);
        args.putString(ARG_PARAM2, param2);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            mParam1 = getArguments().getInt(ARG_PARAM1);
            dataMonth = getArguments().getString(ARG_PARAM2);
        }
        db = FirebaseFirestore.getInstance();
        user = FirebaseAuth.getInstance().getCurrentUser();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_transaction_list, container, false);
        ButterKnife.bind(this, view);


        recyclerView.setHasFixedSize(true);
        LinearLayoutManager stickyHeaderLayoutManager = new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, false);

        recyclerView.setLayoutManager(stickyHeaderLayoutManager);

        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                if (dy > 0) {
                    fabButtonVisibilityListener.hideFabButton();
                } else {
                    fabButtonVisibilityListener.showFabButton();
                }
            }
        });


        initDataLoading(dataMonth);

        return view;
    }

    private void initDataLoading(String dataMonth) {
        getTransactions(dataMonth);
    }

    @OnClick(R.id.btnTryAgain)
    public void tryAgain() {
        initDataLoading(dataMonth);
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        try {
            fabButtonVisibilityListener = (FabButtonVisibilityListener) context;
        } catch (ClassCastException e) {
            throw new ClassCastException(context.toString() + " must implement FabButtonVisibilityListener");
        }
    }

    void getTransactions(String dataMonth) {
        showProgressBar();
        long[] dateIntervals = DateUtils.getDateIntervals(dataMonth);
        Query query = db.collection("users")
                .document(user.getUid())
                .collection("transactions")
                .whereGreaterThanOrEqualTo("date", dateIntervals[0])
                .whereLessThanOrEqualTo("date", dateIntervals[1])
                .orderBy("date", Query.Direction.DESCENDING);

        registration = query.addSnapshotListener(
                (queryDocumentSnapshots, e) -> {

                    if (e != null) {
                        Log.e(TAG, "Listen failed.", e);
                        return;
                    }

                    List<Transaction> transactions = new ArrayList<>();
                    for (QueryDocumentSnapshot doc : queryDocumentSnapshots) {
                        transactions.add(doc.toObject(Transaction.class));
                    }

                    populateResumeInfo(transactions);
                    updateWidget(transactions);

                    if (transactions.isEmpty()) {
                        showNoData();
                        return;
                    }

                    TransactionsAdapter transactionsAdapter = new TransactionsAdapter(getContext(), transactions, this::onTransactionClick);
                    recyclerView.setAdapter(transactionsAdapter);
                    showData();


                });

    }

    private void updateWidget(List<Transaction> transactions) {
        PreferencesManager preferences = PreferencesManager.getInstance(getContext().getSharedPreferences(LAUNCH_PREF, MODE_PRIVATE));

        FirebaseDatabase database = FirebaseDatabase.getInstance();
        Task<Void> task = database.getReference("widget").child(user.getUid()).child("transactions").setValue(getLastFiveTransactions(transactions));

        task.addOnSuccessListener(aVoid -> TransactionViewService.startActionUpdateViewTrsansaction(getContext())).addOnFailureListener(e -> {
            Log.e(TAG, "Saving failed.", e);
            preferences.setTransactions(getLastFiveTransactions(transactions));
            TransactionViewService.startActionUpdateViewTrsansaction(getContext());
        });



    }

    private String getLastFiveTransactions(List<Transaction> transactions) {
        StringBuilder builder = new StringBuilder();
        int count = 0;
        for (Transaction transaction : transactions) {
            if (count < 5) {
                if (DateUtils.getDataDayMonth(transaction.getDate()).equals(DateUtils.getDataDayMonth(new Date().getTime()))) {
                    builder.append("\n")
                            .append(transaction.getCategory())
                            .append(" ")
                            .append(transaction.getType().equals("Income") ? " +" + NumberUtils.getFormattedAmount(transaction.getAmount()) : " -" + NumberUtils.getFormattedAmount(transaction.getAmount()));
                }
            }
            count++;
        }
        return builder.toString();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        registration.remove();
    }

    void showProgressBar() {
        progressBar.setVisibility(View.VISIBLE);
        noDataLayout.setVisibility(View.INVISIBLE);
        noInternetLayout.setVisibility(View.INVISIBLE);
        recyclerView.setVisibility(View.INVISIBLE);
    }

    void showNoInternet() {
        progressBar.setVisibility(View.INVISIBLE);
        noDataLayout.setVisibility(View.INVISIBLE);
        noInternetLayout.setVisibility(View.VISIBLE);
        recyclerView.setVisibility(View.INVISIBLE);
    }

    void showNoData() {
        progressBar.setVisibility(View.INVISIBLE);
        noDataLayout.setVisibility(View.VISIBLE);
        noInternetLayout.setVisibility(View.INVISIBLE);
        recyclerView.setVisibility(View.INVISIBLE);
    }

    void showData() {
        progressBar.setVisibility(View.INVISIBLE);
        noDataLayout.setVisibility(View.INVISIBLE);
        noInternetLayout.setVisibility(View.INVISIBLE);
        recyclerView.setVisibility(View.VISIBLE);
    }

    @Override
    public void onTransactionClick(Transaction transaction) {
        Intent intent = new Intent(getActivity(), DetailsActivity.class);
        intent.putExtra(DetailsActivity.TRANSACTION, transaction);
        startActivity(intent);
    }


    public void openReportActivity(List<Transaction> transactions) {
        Intent intent = new Intent(getActivity(), ReportActivity.class);
        intent.putParcelableArrayListExtra("transactions", new ArrayList<>(transactions));
        startActivity(intent);
    }

    private void populateResumeInfo(List<Transaction> transactions) {
        double totalIncome = getTotalMonthIncome(transactions);
        double totalExpense = getTotalMonthExpense(transactions);

        textViewExpense.setText(String.format("-%s", NumberUtils.getFormattedAmount(totalExpense)));
        textViewExpense.setTextColor(ContextCompat.getColor(getContext(), R.color.colorExpense));
        textViewIncome.setText(String.format("+%s", NumberUtils.getFormattedAmount(totalIncome)));
        textViewIncome.setTextColor(ContextCompat.getColor(getContext(), R.color.colorIncome));
        textViewAvailable.setText(NumberUtils.getFormattedAmount(totalIncome - totalExpense));

        resumeCardView.setOnClickListener(view -> openReportActivity(transactions));
    }

    private double getTotalMonthExpense(List<Transaction> transactions) {
        double expense = 0;
        for (Transaction transaction : transactions) {
            if (transaction.getType().equals("Expense")) {
                expense = expense + transaction.getAmount();
            }
        }
        return expense;
    }

    private double getTotalMonthIncome(List<Transaction> transactions) {
        double income = 0;
        for (Transaction transaction : transactions) {
            if (transaction.getType().equals("Income")) {
                income = income + transaction.getAmount();
            }
        }
        return income;
    }

    public interface FabButtonVisibilityListener {
        void hideFabButton();

        void showFabButton();
    }

}

File: app/src/main/java/com/udacity/maluleque/meutako/TransactionListFragment.kt
package com.udacity.maluleque.meutako

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.TextView
import androidx.cardview.widget.CardView
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView
import butterknife.ButterKnife
import butterknife.OnClick
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.firestore.Query
import com.udacity.maluleque.meutako.adapters.TransactionsAdapter
import com.udacity.maluleque.meutako.adapters.TransactionsAdapter.OnTransactionClickListener
import com.udacity.maluleque.meutako.model.Transaction
import com.udacity.maluleque.meutako.preferences.PreferencesManager
import com.udacity.maluleque.meutako.utils.DateUtils
import com.udacity.maluleque.meutako.utils.NumberUtils
import com.udacity.maluleque.meutako.utils.Status
import com.udacity.maluleque.meutako.viewmodel.TransactionViewModel
import com.udacity.maluleque.meutako.widget.TransactionViewService
import java.util.*
import javax.inject.Inject

class TransactionListFragment : Fragment(), OnTransactionClickListener {

    @BindView(R.id.recyclerViewTransactions)
    lateinit var recyclerView: RecyclerView

    @BindView(R.id.noDataLayout)
    lateinit var noDataLayout: LinearLayout

    @BindView(R.id.noInternetLayout)
    lateinit var noInternetLayout: LinearLayout

    @BindView(R.id.progressBar)
    lateinit var progressBar: ProgressBar

    @BindView(R.id.textViewAvailableAmount)
    lateinit var textViewAvailable: TextView

    @BindView(R.id.textViewIncome)
    lateinit var textViewIncome: TextView

    @BindView(R.id.textViewExpense)
    lateinit var textViewExpense: TextView

    @BindView(R.id.resumeCardView)
    lateinit var resumeCardView: CardView

    private var mParam1 = 0
    private var dataMonth: String? = null
    private var user: FirebaseUser? = null
    private var fabButtonVisibilityListener: FabButtonVisibilityListener? = null
    private lateinit var transactionViewModel: TransactionViewModel

    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            mParam1 = requireArguments().getInt(ARG_PARAM1)
            dataMonth = requireArguments().getString(ARG_PARAM2)
        }
        user = FirebaseAuth.getInstance().currentUser

        (activity?.application as App).appComponent.inject(this)

        transactionViewModel = ViewModelProvider(this, viewModelFactory).get(TransactionViewModel::class.java)

    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_transaction_list, container, false)
        ButterKnife.bind(this, view)
        recyclerView.setHasFixedSize(true)
        val stickyHeaderLayoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
        recyclerView.layoutManager = stickyHeaderLayoutManager
        recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                if (dy > 0) {
                    fabButtonVisibilityListener!!.hideFabButton()
                } else {
                    fabButtonVisibilityListener!!.showFabButton()
                }
            }
        })
        initDataLoading(dataMonth)
        return view
    }

    private fun initDataLoading(dataMonth: String?) {
        getTransactions(dataMonth)
    }

    @OnClick(R.id.btnTryAgain)
    fun tryAgain() {
        initDataLoading(dataMonth)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        fabButtonVisibilityListener = try {
            context as FabButtonVisibilityListener
        } catch (e: ClassCastException) {
            throw ClassCastException("$context must implement FabButtonVisibilityListener")
        }
    }

    fun getTransactions(dataMonth: String?) {
        showProgressBar()
        val dateIntervals = DateUtils.getDateIntervals(dataMonth)

        transactionViewModel.getTransactions(user!!.uid, dateIntervals[0], dateIntervals[1], Query.Direction.DESCENDING)
                .observe(viewLifecycleOwner, androidx.lifecycle.Observer {
                    when (it.status) {
                        Status.LOADING -> {
                            showProgressBar()
                        }
                        Status.SUCCESS -> {
                            val transactions = it.data
                            if (transactions.isNotEmpty()) {
                                populateResumeInfo(transactions)
                                updateWidget(transactions)
                                val transactionsAdapter = TransactionsAdapter(requireContext(), transactions, this)
                                recyclerView!!.adapter = transactionsAdapter
                                showData()
                            } else {
                                showNoData()
                            }
                        }
                        Status.ERROR -> {
                            showNoData()
                        }
                    }
                })

        /* registration = query.addSnapshotListener { queryDocumentSnapshots: QuerySnapshot?, e: FirebaseFirestoreException? ->
             if (e != null) {
                 Log.e(TAG, "Listen failed.", e)
                 return@addSnapshotListener
             }
             val transactions: MutableList<Transaction> = ArrayList()
             for (doc in queryDocumentSnapshots!!) {
                 transactions.add(doc.toObject(Transaction::class.java))
             }
             populateResumeInfo(transactions)
             updateWidget(transactions)
             if (transactions.isEmpty()) {
                 showNoData()
                 return@addSnapshotListener
             }
             val transactionsAdapter = TransactionsAdapter(requireContext(), transactions, this)
             recyclerView!!.adapter = transactionsAdapter
             showData()
         }*/
    }

    private fun updateWidget(transactions: List<Transaction>) {
        val preferences = PreferencesManager.getInstance(requireContext().getSharedPreferences(LAUNCH_PREF, Context.MODE_PRIVATE))
        val database = FirebaseDatabase.getInstance()
        val task = database.getReference("widget").child(user!!.uid).child("transactions").setValue(getLastFiveTransactions(transactions))
        task.addOnSuccessListener { aVoid: Void? -> TransactionViewService.startActionUpdateViewTrsansaction(context) }.addOnFailureListener { e: Exception? ->
            Log.e(TAG, "Saving failed.", e)
            preferences!!.transactions = getLastFiveTransactions(transactions)
            TransactionViewService.startActionUpdateViewTrsansaction(context)
        }
    }

    private fun getLastFiveTransactions(transactions: List<Transaction>): String {
        val builder = StringBuilder()
        var count = 0
        for (transaction in transactions) {
            if (count < 5) {
                if (DateUtils.getDataDayMonth(transaction.date) == DateUtils.getDataDayMonth(Date().time)) {
                    builder.append("\n")
                            .append(transaction.category)
                            .append(" ")
                            .append(if (transaction.type == "Income") " +" + NumberUtils.getFormattedAmount(transaction.amount) else " -" + NumberUtils.getFormattedAmount(transaction.amount))
                }
            }
            count++
        }
        return builder.toString()
    }

    fun showProgressBar() {
        progressBar.visibility = View.VISIBLE
        noDataLayout.visibility = View.INVISIBLE
        noInternetLayout.visibility = View.INVISIBLE
        recyclerView.visibility = View.INVISIBLE
    }

    fun showNoInternet() {
        progressBar.visibility = View.INVISIBLE
        noDataLayout.visibility = View.INVISIBLE
        noInternetLayout.visibility = View.VISIBLE
        recyclerView.visibility = View.INVISIBLE
    }

    fun showNoData() {
        progressBar.visibility = View.INVISIBLE
        noDataLayout.visibility = View.VISIBLE
        noInternetLayout.visibility = View.INVISIBLE
        recyclerView.visibility = View.INVISIBLE
    }

    fun showData() {
        progressBar.visibility = View.INVISIBLE
        noDataLayout.visibility = View.INVISIBLE
        noInternetLayout.visibility = View.INVISIBLE
        recyclerView.visibility = View.VISIBLE
    }

    override fun onTransactionClick(transaction: Transaction?) {
        val intent = Intent(activity, DetailsActivity::class.java)
        intent.putExtra(DetailsActivity.Companion.TRANSACTION, transaction)
        startActivity(intent)
    }

    fun openReportActivity(transactions: List<Transaction>?) {
        val intent = Intent(activity, ReportActivity::class.java)
        intent.putParcelableArrayListExtra("transactions", ArrayList(transactions!!))
        startActivity(intent)
    }

    private fun populateResumeInfo(transactions: List<Transaction>) {
        val totalIncome = getTotalMonthIncome(transactions)
        val totalExpense = getTotalMonthExpense(transactions)
        textViewExpense.text = String.format("-%s", NumberUtils.getFormattedAmount(totalExpense))
        textViewExpense.setTextColor(ContextCompat.getColor(requireContext(), R.color.colorExpense))
        textViewIncome.text = String.format("+%s", NumberUtils.getFormattedAmount(totalIncome))
        textViewIncome.setTextColor(ContextCompat.getColor(requireContext(), R.color.colorIncome))
        textViewAvailable.text = NumberUtils.getFormattedAmount(totalIncome - totalExpense)
        resumeCardView.setOnClickListener { view: View? -> openReportActivity(transactions) }
    }

    private fun getTotalMonthExpense(transactions: List<Transaction>): Double {
        var expense = 0.0
        for (transaction in transactions) {
            if (transaction.type == "Expense") {
                expense = expense + transaction.amount
            }
        }
        return expense
    }

    private fun getTotalMonthIncome(transactions: List<Transaction>): Double {
        var income = 0.0
        for (transaction in transactions) {
            if (transaction.type == "Income") {
                income = income + transaction.amount
            }
        }
        return income
    }

    interface FabButtonVisibilityListener {
        fun hideFabButton()
        fun showFabButton()
    }

    companion object {
        private const val ARG_PARAM1 = "param1"
        private const val ARG_PARAM2 = "param2"
        private const val TAG = "TransactionListFragment"
        private const val LAUNCH_PREF = "launch-prefs"

        @JvmStatic
        fun newInstance(param1: Int, param2: String?): TransactionListFragment {
            val fragment = TransactionListFragment()
            val args = Bundle()
            args.putInt(ARG_PARAM1, param1)
            args.putString(ARG_PARAM2, param2)
            fragment.arguments = args
            return fragment
        }
    }
}